### DISCUSSION TITLE
Fonts and reproducibility

### DISCUSSION THREAD

**Discussion by ask6155** [tectonic-typesetting/tectonic#1228] [Q&A]
Hello,

I tried messing around with tectonic and here are the things I have questions about.

1. If i use the command `\newfontfamily{font}` I can access font installed on my system. It warns that using them will make the build system not reproducible. How do I fix this?
2. The tectonic book also says it has the ability to download fonts. However, I have not been able to do so. I have tried to import fonts from the ctan archive that are not on my system but it just says font not found on the system.
3. I used to use lualatex and I used the package `pdfx` to comply with `pdf/a-3u`. However, when I use `pdfx` with tectonic it gives me the error that `error: pdfx.sty:1285: Package pdfx Error: CreationDate is not properly supported;
 PDF validation may fail..` How do I fix this?

**Comment by vlasakm** [comment]
Hello!

> If i use the command \newfontfamily{font} I can access font installed on my system. It warns that using them will make the build system not reproducible. How do I fix this?

I think the right fix is to use fonts from Tectonic bundles, as that will make your document reproducible and make the warnings disappear. Which brings us to:

> The tectonic book also says it has the ability to download fonts. However, I have not been able to do so. I have tried to import fonts from the ctan archive that are not on my system but it just says font not found on the system.

Tectonic can download any file from CTAN. Or more precisely, Tectonic has a concept of bundles which are snapshots of the yearly releases of TeX Live (like 2022, 2023, 2024), which include almost everything on CTAN (only excluding some packages with licensing issues).

This includes fonts as well. However, with fonts it's a bit more tricky as there are actually two ways it is possible to ask "TeX" for a font: by _font file name_ (e.g. `myfont.ttf`) or _font names_ (e.g. `My Font`). File names work without problem, as Tectonic just directly finds the file in the bundle. Font names are more problematic, as for them the `fontconfig` library is used, and there is currently no nice way to make `fontconfig` look into Tectonic bundles (as it's just a list of files that are available over the internet). That's why even when a font is available in CTAN (and hence Tectonic bundle) you usually still want to ensure you use font file names to lookup the fonts, instead of font names.

I think using font file names is a better approach for more reasons other than the fact that font names are not supported well in Tectonic.

I have an example and more details in https://github.com/tectonic-typesetting/tectonic/issues/965#issuecomment-1317865762. The issue about  not being able to find fonts by font names in bundles is https://github.com/tectonic-typesetting/tectonic/issues/9.

> I used to use lualatex and I used the package `pdfx` to comply with `pdf/a-3u`. However, when I use `pdfx` with tectonic it gives me the error that `error: pdfx.sty:1285: Package pdfx Error: CreationDate is not properly supported;  PDF validation may fail..` How do I fix this?

`pdfx` not completely supported with Tectonic, the issue is tracked in https://github.com/tectonic-typesetting/tectonic/issues/838. IIRC we fixed some problems in Tectonic, but the `\creationdate` problem is a problem in the `pdfx` package itself.

In case of `CreationDate`, the following should work as a workaround:

```
\let\pdfcreationdate=\creationdate
```

It actually seems that the `\creationdate` problem was [resolved](https://git.texlive.info/texlive/commit/?id=1102f1b8bdada5b85eb1bf1831eaf3f8bee37506) [in a recent release (2024-07-02)](https://ctan.org/ctan-ann/id/mailman.77.1719900776.3550.ctan-ann@ctan.org). Because of the yearly basis of TeX Live (release at around March each year) and Tectonic somewhat lagging behind, this change may take a bit of time before it reaches what Tectonic would be using by default, so the workaround may be needed until then.

If you have more questions, please include a [MWE](https://tex.meta.stackexchange.com/questions/228/ive-just-been-asked-to-write-a-minimal-working-example-mwe-what-is-that). MWE gives better chance of reproducing problems and also gives a chance to suggest a fix for your particular use case. In this case for example, I don't know what exact fonts you are trying to use, and thus don't know if you need to adjust the way you use the `fontspec` package, or whether you use some CTAN package that loads the font and the package uses font names.

**Comment by ask6155** [comment]
Thank you for your reply
In this example:
```latex
\documentclass{article}

\usepackage{fontspec}
\usepackage[margin=0.5in]{geometry}

\newfontfamily\mono{FreeMono.otf}
\setlength{\parindent}{0ex}

\title{My Title}
\begin{document}
\section{Testing}

Hello.

\mono This is in Monospace.
```
The font `FreeMono` is pulled from the bundle. When I remove the `.otf` extension it pulls it from my system.
But what if a font is not in TeX live? How would I supply such a font? (In a reproducible manner)
Can I include it in my working directory?

I tried `\let\pdfcreationdate=\creationdate` and I could compile a document. Horray!

**Reply by vlasakm** [reply]
> The font FreeMono is pulled from the bundle. When I remove the .otf extension it pulls it from my system.

That sounds accurate to me - `FreeMono` is font  _name_ lookup, `FreeMono.otf` is font _file name_ lookup, both going with very different internal code paths.

Note that like in https://github.com/tectonic-typesetting/tectonic/issues/965#issuecomment-1317865762, something more elaborate would be required to properly set the fonts including all the styles (bold, etc.) when using file names, e.g.:

```latex
\documentclass{article}

\usepackage{fontspec}

\setromanfont{SourceSerifPro}[
	Extension      = .otf,
	UprightFont    = *-Regular,
	BoldFont       = *-Bold,
	ItalicFont     = *-RegularIt,
	BoldItalicFont = *-BoldIt]

\setsansfont{NotoSans}[
	Extension      = .ttf,
	UprightFont    = *-Regular,
	BoldFont       = *-Bold,
	ItalicFont     = *-Italic,
	BoldItalicFont = *-BoldItalic]

\setmonofont{SourceCodePro}[
	Extension      = .otf,
	UprightFont    = *-Regular,
	BoldFont       = *-Bold,
	ItalicFont     = *-RegularIt,
	BoldItalicFont = *-BoldIt]


\begin{document}

\textrm{a} \textsf{b} \texttt{c}

\end{document}
```

> But what if a font is not in TeX live? How would I supply such a font? (In a reproducible manner)
> Can I include it in my working directory?

I believe so, though I have to admit I didn't need to do that with Tectonic yet. If it doesn't work I believe it is a bug, as having the font file with the source code is the most reproducible way, other than the Tectonic bundles, certainly much better than using system fonts.

Though recurring problem is: what if you have multiple documents, then having a copy of font per document is not convenient, if you don't have all the documents in one directory. I believe there has been recent progress in improving Tectonic's search paths, and that this actually now works much better.

**Reply by vlasakm** [reply]
> I tried \let\pdfcreationdate=\creationdate and I could compile a document. Horray!

Glad it worked. I am not sure that's everything to make `pdfx` work really properly. But if you find that a validator complains about Tectonic produced files, but not XeLaTeX ones, open an issue.

**Comment by ask6155** [comment]
Another problem I ran into. I tried to run pgfplots with the externalize feature:
```latex
\documentclass{article}

\usepackage{pgfplots}
\usepackage[margin=0.5in]{geometry}
\pgfplotsset{width=10cm,compat=1.18}

\usepgfplotslibrary{external}
\tikzexternalize

\title{My Title}
\begin{document}

\section{Testing}

\begin{tikzpicture}
    \draw (0,0) -- (1,1);
\end{tikzpicture}
```

I get the error
```
warning: _preamble.tex:23: runsystem(xelatex -halt-on-error -interaction=batchmode -jobname "default-figure0" "\def\tikzexternalrealjob{default}\input{default}")...disabled.
error: _preamble.tex:23: Package tikz Error: Sorry, the system call 'xelatex -halt-on-error -interaction=batchmode -jobname "default-figure0" "\def\tikzexternalrealjob{default}\input{default}"' did NOT result in a usable output file 'default-figure0' (expected one of .pdf:.jpg:.jpeg:.png:.bmp:). Please verify that you have enabled system calls. For pdflatex, this is 'pdflatex -shell-escape'. Sometimes it is also named 'write 18' or something like that. Or maybe the command simply failed? Error messages can be found in 'default-figure0.log'. If you continue now, I'll try to typeset the picture.

```

**Reply by vlasakm** [reply]
TikZ externalize feature in the background launches externall processes, which is forbidden by default in all TeX engines. You need to enable the shell escape option, `-shell-escape` in usual TeX engines, `-Z shell-escape` in Tectonic.

This is also discussed in https://github.com/tectonic-typesetting/tectonic/issues/852. The full command line should also likely include `-Z shell-escape-cwd=./`, so that the correct working directory gets used for the command.

However, there seems to be a second issue, also already known https://github.com/tectonic-typesetting/tectonic/issues/852#issuecomment-1317300594, i.e. TikZ tries to process the external figure with `xelatex` command, but that is wrong, since we need it to execute `tectonic` (and with a different command line). This means that if you have TeX Live installed (and in `$PATH` it might work for you "accidentally", but if you don't, it won't work. This seems like something we should work around in the Tectonic bundle.

**Comment by ask6155** [comment]
I included the font in my source as stated in #1095 and sure enough it worked!
However, I couldn't test out shell escape for Tikz externalize as `-Z` is not supported with `-X`?
I had to use V1 interface and then the command ran but didn't compile as TeX Live was not installed
```
Running TeX ...
note: running shell command: `xelatex -shell-escape -halt-on-error -interaction=batchmode -jobname "_preamble-figure0" "\def\tikzexternalrealjob{_preamble}\input{_preamble}"`
sh: line 1: xelatex: command not found
warning: command exited with error code 127
error: failed to execute the shell-escape command "xelatex -shell-escape -halt-on-error -interaction=batchmode -jobname "_preamble-figure0" "\def\tikzexternalrealjob{_preamble}\input{_preamble}"": execution of the request failed
error: _preamble.tex:49: Package tikz Error: Sorry, the system call 'xelatex -shell-escape -halt-on-error -interaction=batchmode -jobname "_preamble-figure0" "\def\tikzexternalrealjob{_preamble}\input{_preamble}"' did NOT result in a usable output file '_preamble-figure0' (expected one of .pdf:.jpg:.jpeg:.png:.bmp:). Please verify that you have enabled system calls. For pdflatex, this is 'pdflatex -shell-escape'. Sometimes it is also named 'write 18' or something like that. Or maybe the command simply failed? Error messages can be found in '_preamble-figure0.log'. If you continue now, I'll try to typeset the picture.

```

**Reply by kpym** [reply]
you can check my answer in [#852](https://github.com/tectonic-typesetting/tectonic/issues/852#issuecomment-2430109576).

**Comment by ask6155** [comment]
This was a fruitful discussion (atleast for me 😁) Now i can switch over to tectonic fulltime I think. Thank you for your help guys 🙏
